daemon/cmd/daemon.go

```diff
@@ -276,7 +277,55 @@ func restoreCiliumHostIPs(ipv6 bool, fromK8s net.IP) {
 		fromFS = node.GetInternalIPv4Router()
 	}
 
-	node.RestoreHostIPs(ipv6, fromK8s, fromFS, cidr)
+	restoredIP := node.RestoreHostIPs(ipv6, fromK8s, fromFS, cidr)
+	if err := removeOldRouterState(restoredIP); err != nil {
+		log.WithError(err).Warnf(
+			"Failed to remove old router IPs (restored IP: %v) from cilium_host. Manual intervention is required to remove all other old IPs.",
+			restoredIP,
+		)
+	}
+}
+
+// removeOldRouterState will try to ensure that the only IP assigned to the
+// `cilium_host` interface is the given restored IP. If the given IP is nil,
+// then it attempts to clear all IPs from the interface.
+func removeOldRouterState(restoredIP net.IP) error {
+	l, err := netlink.LinkByName(defaults.HostDevice)
+	if err != nil {
+		return err
+	}
+
+	family := netlink.FAMILY_V6
+	if restoredIP.To4() != nil {
+		family = netlink.FAMILY_V4
+	}
+	addrs, err := netlink.AddrList(l, family)
+	if err != nil {
+		return err
+	}
+	if len(addrs) > 1 {
+		log.Info("More than one router IP was found on the cilium_host device after restoration, cleaning up old router IPs.")
+
+		var errs []error
+		for _, a := range addrs {
+			if restoredIP == nil {
+				goto deleteIP
+			}
+			if restoredIP.Equal(a.IP) {
+				continue
+			}
+
+		deleteIP:
+			if err := netlink.AddrDel(l, &a); err != nil {
+				errs = append(errs, err)
```

Comment:
This is what I was trying to get at in the [earlier thread](https://github.com/cilium/cilium/pull/17762#discussion_r741476983):

```suggestion
				errs = append(errs, fmt.Errorf("cannot remove IP %s from device: %w", a, err))
```